### PR TITLE
fix: issue with TMS conversion when using S3

### PR DESCRIPTION
**Pull request by shimoncohen**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✔                                                                        |
| New feature     | ✖                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✔                                                                       |

### CODE REVIEW

File: Src/Batching/TIle.cs [hunk 1]
```diff
@@ -28,6 +28,11 @@ public Tile(int z, int x, int y, string blob, int blobSize)
             this.BlobSize = blobSize;
         }
 
+        public bool HasCoords(int z, int x, int y)
+        {
+            return z == this.X && x == this.X && y == this.Y;
```
Comment (cl-shaharshavit-rnd) [inline Src/Batching/TIle.cs]: z == x?

File: Src/DataTypes/Coord.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+        public int y;
+
+        public Coord(int z, int x, int y)
+        {
+            this.z = z;
+            this.x = x;
+            this.y = y;
+        }
+    }
+}
```
Comment (cl-shaharshavit-rnd) [inline Src/DataTypes/Coord.cs:16]: line break

File: Src/DataTypes/Gpkg.cs [hunk 3]
```diff
@@ -150,6 +150,7 @@ public override void Cleanup()
 
         public override bool Exists()
         {
+            Console.WriteLine($"Checking if exists, gpkg: {this.path}");
```
Comment (cl-shaharshavit-rnd) [inline Src/DataTypes/Gpkg.cs:153]: should use logger if this will be add to service ( log4net maybe ?)

File: Src/DataTypes/S3.cs [hunk 4]
```diff
@@ -35,6 +35,19 @@ public S3(string serviceUrl, string bucket, string path, int batchSize) : base(D
             client = new AmazonS3Client(credentials, config);
         }
 
+        private Coord FromKey(string key)
+        {
+            string subPath = key.Remove(0, this.path.Length);
```
Comment (cl-shaharshavit-rnd) [inline Src/DataTypes/S3.cs:40]: what if the path is relative? (without .)
I think split have option to skip empty parts instead (if not linq can be used)

File: Src/DataTypes/S3.cs [hunk 5]
```diff
@@ -35,6 +35,19 @@ public S3(string serviceUrl, string bucket, string path, int batchSize) : base(D
...
+            string[] parts = subPath.Split('/');
+
+            string[] last = parts[2].Split('.');
+            int z = int.Parse(parts[0]);
+            int x = int.Parse(parts[1]);
+            int y = int.Parse(last[0]);
```
Comment (cl-shaharshavit-rnd) [inline Src/DataTypes/S3.cs:46]: z and y are the same?

File: Src/Utils/GeoUtils.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+            // Convert to and from TMS
+            return (1 << z) - y - 1;
+        }
+
+        public static int convertTMS(Tile tile)
+        {
+            return convertTMS(tile.Z, tile.Y);
+        }
+    }
+}
```
Comment (cl-shaharshavit-rnd) [inline Src/Utils/GeoUtils.cs:18]: line break

File: Src/Utils/S3Utils.cs [hunk 7]
```diff
@@ -39,21 +40,37 @@ public static string GetImageHex(AmazonS3Client client, string bucket, string ke
             }
         }
 
-        public static Tile GetTile(AmazonS3Client client, string bucket, string path, int z, int x, int y)
+        public static Tile GetTileTMS(AmazonS3Client client, string bucket, string path, int z, int x, int y)
         {
-            Tile correspondingTile = null;
-            string key = $"{path}{z}/{z}/{y}.png";
+            Tile tile = null;
+            string key = $"{path}{z}/{x}/{y}.png";
```
Comment (cl-shaharshavit-rnd) [inline Src/Utils/S3Utils.cs:46]: missing / before z ?

File: Src/Utils/S3Utils.cs [hunk 8]
```diff
@@ -39,21 +40,37 @@ public static string GetImageHex(AmazonS3Client client, string bucket, string ke
...
-            return correspondingTile;
+            return tile;
+        }
+
+        public static Tile GetTileTMS(AmazonS3Client client, string bucket, string path, Coord coord)
+        {
+            return GetTileTMS(client, bucket, path, coord.z, coord.x, coord.y);
+        }
+
+        public static Tile GetTile(AmazonS3Client client, string bucket, string path, int z, int x, int y)
```
Comment (cl-shaharshavit-rnd) [inline Src/Utils/S3Utils.cs:63]: why not use Coord object too?

**Review by cl-shaharshavit-rnd — CHANGES_REQUESTED** [review]

File: Src/Batching/TIle.cs (hunk 1 above)
Comment (shimoncohen) [reply Src/Batching/TIle.cs]: Fixed

Comment (shimoncohen) [reply Src/Batching/TIle.cs]: Fixed

File: Src/DataTypes/Coord.cs (hunk 2 above)
Comment (shimoncohen) [reply Src/DataTypes/Coord.cs:16]: Fixed

File: Src/DataTypes/Gpkg.cs (hunk 3 above)
Comment (shimoncohen) [reply Src/DataTypes/Gpkg.cs:153]: When transforming to a worker, this will be part of the proccess.

File: Src/DataTypes/S3.cs (hunk 4 above)
Comment (shimoncohen) [reply Src/DataTypes/S3.cs:40]: This is a S3 key, the path cannot be relative.
It is retrieved from S3 itself and is not user input.

File: Src/DataTypes/S3.cs (hunk 5 above)
Comment (shimoncohen) [reply Src/DataTypes/S3.cs:46]: They are not, please look closely.

File: Src/Utils/GeoUtils.cs (hunk 6 above)
Comment (shimoncohen) [reply Src/Utils/GeoUtils.cs:18]: Added

File: Src/Utils/S3Utils.cs (hunk 8 above)
Comment (shimoncohen) [reply Src/Utils/S3Utils.cs:63]: There was no use for it in this context.
The places from which it was called already had a tile at hand or the values of z,x,y.

File: Src/Utils/S3Utils.cs (hunk 7 above)
Comment (shimoncohen) [reply Src/Utils/S3Utils.cs:46]: Fixed

File: Src/Utils/PathUtils.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
+namespace GpkgMerger.Src.Utils
+{
+    public static class PathUtils
+    {
+        public static string RemoveTrailingSlash(string path)
```
Comment (cl-shaharshavit-rnd) [inline Src/Utils/PathUtils.cs:5]: why not use trim/trimEnd ('/')?

**Review by cl-shaharshavit-rnd — APPROVED** [review]

File: Src/Utils/PathUtils.cs (hunk 9 above)
Comment (shimoncohen) [reply Src/Utils/PathUtils.cs:5]: Changed

**Review by cl-shaharshavit-rnd — APPROVED** [review]

**Pull request merged**